service/src/IOUringImp.cpp

```
@@ -0,0 +1,104 @@
+        m_result_destinations.push_back(destination);
+        io_uring_sqe_set_data(sqe, destination.get());
+    }
+
+    void IOUringImp::submit()
+    {
+        int submitted_operations = io_uring_submit(&m_ring);
+        struct io_uring_cqe *cqe;
+
+        for (int operation = 0; operation < submitted_operations; ++operation) {
+            int ret = io_uring_wait_cqe(&m_ring, &cqe);
+            if (ret < 0) {
+                throw Exception("Failed to get a completion event from IO uring",
+                                -ret, __FILE__, __LINE__);
+            }
+
+            int* result_destination = static_cast<int*>(io_uring_cqe_get_data(cqe));
+            if (result_destination) {
+                // The caller of prep_...() for this operation wants to
+                // know the return value of the operation, so write it back.
+                *result_destination = cqe->res;
+            }
+            io_uring_cqe_seen(&m_ring, cqe);
+        }
+
+        // We're done writing batch operation results, so we don't need to
+        // track the destination pointers any more.
+        m_result_destinations.clear();
+    }
+
+    void IOUringImp::prep_read(std::shared_ptr<int> ret, int fd, void *buf,
+                               unsigned nbytes, off_t offset)
+    {
+        auto sqe = get_sqe_or_throw();
+        set_sqe_return_destination(sqe, ret);
+        // Available since Linux 5.10
+        io_uring_prep_read(sqe, fd, buf, nbytes, offset);
+    }
+
+    void IOUringImp::prep_write(std::shared_ptr<int> ret, int fd, const void *buf,
+                                unsigned nbytes, off_t offset)
+    {
+        auto sqe = get_sqe_or_throw();
+        set_sqe_return_destination(sqe, ret);
+        // Available since Linux 5.10
+        io_uring_prep_write(sqe, fd, buf, nbytes, offset);
+    }
+
+    std::shared_ptr<IOUring> make_io_uring_imp(unsigned entries)
+    {
+        return std::make_shared<IOUringImp>(entries);
```

We could also probe for our needed features and throw if they (or the probing feature) are absent. The caller of this function could `try { make the uring version } catch (...) { make the fallback version }`